File: epi/Python/list_12.py
# Implement even-odd merge

class Node:
  def __init__(self, data, node):
    self.data = data
    self.next = node
    
  def __str__(self):
    return str(self.data)

  def linkedlist(self):
    runner = self;
    res = ''
    while runner.next != None:
      runner = runner.next
      res += str(runner) + ' '
    return res.strip()


def even_odd_merge(head):
  if (head == None or head.next == None or head.next.next == None):
    return
  else:
    even = head.next
    odd = head.next.next
    runner_even = even
    runner_odd = odd
    while True:
      runner_even.next = runner_odd.next
      runner_even = runner_even.next
      if runner_even.next == None:
        runner_odd.next = None
        break
      runner_odd.next = runner_even.next
      runner_odd = runner_odd.next
      if runner_odd.next == None:
        break
    runner_even.next = odd

x = Node(None, Node(0, Node(1, Node(2, Node(3, Node(4, Node(5, None)))))))
y = Node(None, Node(0, Node(1, Node(2, Node(3, Node(4, None))))))
even_odd_merge(x)
even_odd_merge(y)
print x.linkedlist()
print y.linkedlist()

File: epi/Ruby/list_12.rb
=begin
Implement even-odd merge
Write a function that takes a singly linked list L, and reorders the elements
in L so that new list represents even-odd elements from the original one
alternatively. Your function should use O(1) additional space. The only field
you can change in the node is next.
=end
require "libraries"

def even_odd(head)
  return head if head.nil? || head.next.nil? || head.next.next.nil?
  odd = head.next.next
  even_runner = head.next
  odd_runner = odd
  until odd_runner.nil? || odd_runner.next.nil?
    even_runner.next = odd_runner.next
    odd_runner.next = odd_runner.next.next
    even_runner = even_runner.next
    odd_runner = odd_runner.next
  end
  even_runner.next = odd
  head
end

head = Node.new(nil, Node.new(0, Node.new(1, Node.new(2, Node.new(3, Node.new(4,
Node.new(5, nil)))))))
puts even_odd(head).print_list
